src/migrate/v0.4.md

```diff
@@ -0,0 +1,398 @@
+
+[compat]: ../toolchain/compatibility.md
+
+[#1292]: https://github.com/godot-rust/gdext/pull/1292
+
+
+## Overhauls to core APIs
+
+
+### Argument passing redesign
+
+[#1285], [#1308], [#1310], [#1314]: So far, [`AsArg`][api-asarg] and [`ToGodot`][api-togodot] traits have had very similar roles, namely converting
+Rust types to Godot-compatible types. However, they had only limited interop. For example, it wasn't easily possible to pass user-defined enums to
+`emit()` of signals.
+
+
+#### `ToGodot::Pass` and automatic `AsArg` impls
+
+The `ToGodot` trait has been simplified, replacing the associated type `ToVia<'v>` with `Pass`, which usually has one of two values.
+`AsArg` is then auto-implemented for types that support `ToGodot`, in the following way:
+
+- `Pass = ByValue`: what you typically want. This implements `AsArg<T>` for owned types `T`, e.g. you can pass `i64` directly to Godot.
+- `Pass = ByRef` for specifically optimized types. This implements `AsArg<T>` for references `&T`, e.g. you can pass `&GString` to Godot
+  without cloning. As a user, you typically need to convert to a custom type anyway, so `ByValue` is often easier.
+
+If you follow those, you should never need to implement `AsArg` yourself. `ToGodot` is sufficient, and can often be derived.
+
+**Before (0.3):**
+
+```rust
+impl ToGodot for MyString {
+    type ToVia<'v> = Self::Via;
+
+    fn to_godot(&self) -> Self::ToVia<'_> {
+        GString::from(self)
+    }
+}
+```
+
+**After (0.4):**
+
+```rust
+use godot::meta;
+
+impl ToGodot for MyString {
+    type Pass = meta::ByValue;
+
+    // Returns a new GString here, since we have to create one anyway.
+    fn to_godot(&self) -> Self::Via {
+        GString::from(self)
+    }
+}
+
+impl ToGodot for MyCachedString {
+    type Pass = meta::ByRef;
+
+    // Returns &GString here, since we store a reference.
+    fn to_godot(&self) -> &Self::Via {
+        &self.cached_gstring
+    }
+}
+```
+
+[api-asarg]: https://godot-rust.github.io/docs/gdext/master/godot/meta/trait.AsArg.html
+[api-togodot]: https://godot-rust.github.io/docs/gdext/master/godot/meta/trait.ToGodot.html
+
+
+#### Return type changes
+
+For reference-counted types (`GString`, `Array`, `Dictionary`, `Variant`...), `to_godot()` now returns references.
+A value can be obtained by calling `to_godot_owned()`.
+
+**Before (0.3):**
+
+```rust
+let a: GString = my_value.to_godot();
+```
+
+**After (0.4):**
+
+```rust
+let a: &GString = my_value.to_godot();
+let b: GString = my_value.to_godot_owned();
+```
+
+
+#### Object argument traits
+
+The specialized `AsObjectArg<T>` trait has been consolidated into the more general [`AsArg<Gd<T>>`][api-asarg] trait, unifying the type system.
+It is also capable of expressing optional arguments through `AsArg<Option<Gd<T>>>`.
+
+**Before (0.3):**
+
+```rust
+fn engine_function<T>(obj: impl AsObjectArg<T>) { }
+```
+
+**After (0.4):**
+
+```rust
+fn engine_function<T>(obj: impl AsArg<Gd<T>>) { }
+```
```

Comment:
```suggestion

```rust
#[signal]
fn MySignal(some_node: Gd<Node2D>);

// Argument had to be upcasted.
let arg = MyNode::new_alloc();
sig.emit(&arg.upcast());

let b = Node::new_alloc();
// The type could be inferred, but arguments had to be implicitly upcasted.
let _array: Array<Gd<Node>> = array![&arg.upcast(), &b];
```

**After (0.4):**

```rust
fn MySignal(some_node: Gd<Node2D>);

// Will be implicitly upcasted.
let arg = MyNode::new_alloc();
sig.emit(&arg);

let b = Node::new_alloc();
// Type must be specified, but arguments will be implicitly upcasted.
let _array: Array<Gd<Node>> = array![&arg, &b];

```

```
?

Reply:
Good idea! In 0.3, I could omit the explicit type `: Array<...>` then 🤔